capsules/src/lpm013m126.rs

```diff
@@ -0,0 +1,487 @@
+            height: height as u16,
+        };
+
+        let mut new_state = None;
+        let ret = match self.state.get() {
+            State::Uninitialized => Err(ErrorCode::OFF),
+            State::InitializingPixelMemory | State::InitializingRest => Err(ErrorCode::BUSY),
+            State::Idle(..) => {
+                new_state = Some(State::Idle(frame));
+                Ok(())
+            }
+            State::Writing(..) => {
+                new_state = Some(State::Writing(frame));
+                Ok(())
+            }
+            State::Bug => Err(ErrorCode::FAIL),
+        };
+
+        if let Some(new_state) = new_state {
+            self.state.set(new_state);
+        }
+        // Thankfully, this is the only command that results in the callback,
+        // so there's no danger that this will get attributed
+        // to a command that's not finished yet.
+        self.client.map(|client| client.command_complete(Ok(())));
+        ret
+    }
+
+    fn write(&self, buffer: &'static mut [u8], len: usize) -> Result<(), ErrorCode> {
+        let ret = match self.state.get() {
+            State::Uninitialized => Err(ErrorCode::OFF),
+            State::InitializingPixelMemory | State::InitializingRest => Err(ErrorCode::BUSY),
+            State::Idle(frame) => {
+                self.frame_buffer
+                    .take()
+                    .map_or(Err(ErrorCode::NOMEM), |mut frame_buffer| {
+                        frame_buffer.blit(&buffer[..len], &frame);
+                        let send_buf = FrameBuffer::with_raw_rows(
+                            frame_buffer,
+                            frame.row,
+                            frame.row + frame.height,
+                        );
+                        //send_buf[0] = 67;
+                        //send_buf[2] = 0;
+                        //debug!("{}, {:?}", send_buf.len(), &send_buf[47..73]);
+                        let sent = self.spi.read_write_bytes(send_buf, None, send_buf.len());
+                        let (ret, new_state) = match sent {
+                            Ok(()) => (Ok(()), State::Writing(frame)),
+                            Err((e, buf, _)) => {
+                                self.frame_buffer.replace(FrameBuffer::new(buf));
+                                (Err(e), State::Idle(frame))
+                            }
+                        };
+                        self.state.set(new_state);
+                        ret
+                    })
+            }
+            State::Writing(..) => Err(ErrorCode::BUSY),
+            State::Bug => Err(ErrorCode::FAIL),
+        };
+
+        self.buffer.replace(buffer);
+        match self.state.get() {
+            State::Writing(..) => {}
+            _ => {
+                self.client.map(|client| {
+                    self.buffer
+                        .take()
+                        .map(|buffer| client.write_complete(buffer, ret))
+                });
```

Comment:
Same here, although it's not clear to me why a write_complete callback would be triggered while the driver is still in the `Writing` state.